x-pack/plugins/maps/public/classes/layers/choropleth_layer_wizard/layer_template.tsx

```diff
@@ -0,0 +1,450 @@
+                defaultMessage: 'Select field',
+              })}
+              value={this.state.leftJoinField}
+              onChange={this._onLeftJoinFieldSelect}
+              fields={this.state.leftJoinFields}
+              isClearable={false}
+            />
+          </EuiFormRow>
+        );
+      }
+      return (
+        <>
+          <GeoIndexPatternSelect
+            // @ts-expect-error - avoid wrong "Property 'id' does not exist on type 'never'." compile error
+            value={this.state.leftIndexPattern ? this.state.leftIndexPattern!.id : ''}
+            onChange={this._onLeftIndexPatternChange}
+          />
+          {geoFieldSelect}
+          {joinFieldSelect}
+        </>
+      );
+    } else {
+      let emsFieldSelect;
+      if (this.state.leftEmsFields.length) {
+        let selectedOption;
+        if (this.state.leftJoinField) {
+          selectedOption = this.state.leftEmsFields.find(
+            (option: EuiComboBoxOptionOption<string>) => {
+              return this.state.leftJoinField === option.value;
+            }
+          );
+        }
+        emsFieldSelect = (
+          <EuiFormRow
+            label={i18n.translate('xpack.maps.choropleth.joinFieldLabel', {
+              defaultMessage: 'Join field',
+            })}
+          >
+            <EuiComboBox
+              singleSelection={true}
+              isClearable={false}
+              options={this.state.leftEmsFields}
+              selectedOptions={selectedOption ? [selectedOption] : []}
+              onChange={this._onLeftEmsFieldChange}
+            />
+          </EuiFormRow>
+        );
+      }
+      return (
+        <>
+          <EMSFileSelect value={this.state.leftEmsFileId} onChange={this._onLeftEmsFileChange} />
+          {emsFieldSelect}
+        </>
+      );
+    }
+  }
+
+  _renderLeftPanel() {
+    return (
+      <EuiPanel>
+        <EuiTitle size="xs">
+          <h5>
+            <FormattedMessage
+              id="xpack.maps.choropleth.boundariesLabel"
+              defaultMessage="Boundaries source"
+            />
+          </h5>
+        </EuiTitle>
+
+        <EuiSpacer size="m" />
+
+        <EuiFormRow>
+          <EuiRadioGroup
+            options={BOUNDARIES_OPTIONS}
+            idSelected={this.state.leftSource}
+            onChange={this._onLeftSourceChange}
+          />
+        </EuiFormRow>
+
+        {this._renderLeftSourceForm()}
+      </EuiPanel>
+    );
+  }
+
+  _renderRightPanel() {
+    if (!this._isLeftConfigComplete()) {
+      return null;
+    }
+    const IndexPatternSelect = getIndexPatternSelectComponent();
+
+    let joinFieldSelect;
+    if (this.state.rightTermsFields.length) {
+      joinFieldSelect = (
+        <EuiFormRow
+          label={i18n.translate('xpack.maps.choropleth.joinFieldLabel', {
+            defaultMessage: 'Join field',
+          })}
+        >
+          <SingleFieldSelect
```

Comment:
I would like to leave this optimization off from the initial PR. Seems like it could be error prone and it would be better to implement https://github.com/elastic/kibana/issues/69856 first and then make the recommended optimization part of that effort.